### DISCUSSION TITLE
Mark kwargs as secret

### DISCUSSION THREAD

**Discussion by WilliamDEdwards** [celery/celery#7322] [Q&A]
How do I tell Celery not to store certain keyword arguments to the result backend? Should I pop it myself?

I've set `result_extended` to `True` because I need access to some keyword arguments. However, the kwargs also contains two passwords, which are currently stored as plain text.

**Comment by thedrow** [comment]
You should make a feature request since this is currently not supported.

**Reply by WilliamDEdwards** [reply]
Thanks. I'll try to make a PR.

**Reply by WilliamDEdwards** [reply]
May I ask for your thoughts on the specs of this possible feature?

* Do we want to allow users to specify secret kwargs on a per-task basis by adding an attribute to the `Task` class? I think this would be ideal, but I haven't been able to figure out how to access those attributes in `backends.base.Backend._get_result_meta`.
* Do we want to pop the kwarg completely? If we do, I propose we do it as early as possible. This avoids kwargs leaking if a developer forgets to masquerade the kwargs. On the other hand, popping the kwarg implies it wasn't there at all, which could lead to confusion. But changing its value doesn't seem better. Something like this in `backends.base.Backend._get_result_meta`:

```
@ -480,6 +486,16 @@ class Backend:

         if self.app.conf.find_value_for_key('extended', 'result'):
             if request:
+                # Remove secret kwargs before setting request meta, which includes
+                # kwargs
+                secret_kwargs = self.app.conf.get('result_secret_kwargs', [])
+                for sk in secret_kwargs:
+                    try:
+                        request.kwargs.pop(sk)
+                    except KeyError:
+                        # kwarg may be absent
+                        pass
```

**Reply by stuaxo** [reply]
As you mentioned, having a value is a useful to the developer as they know it was present.

It might be worth using a special class as a sentinel value, to distinguish it from ordinary strings, even if the special class is just equivalent to:

```
class SecretValue:
    pass
```
